airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/PostgresCtidHandler.java

```diff
@@ -110,47 +123,100 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(
+    } else {
+      chunks.add(Pair.of(Ctid.of(lowerBound, startCtid.tuple), Ctid.of(upperBound, 0)));
+      while (upperBound < theoreticalLastPage) {
+        lowerBound = upperBound;
+        upperBound += eachStep;
+        chunks.add(Pair.of(Ctid.of(lowerBound, 0), upperBound > theoreticalLastPage ? null : Ctid.of(upperBound, 0)));
+      }
+    }
+    // The last pair is (x,y) -> null to indicate an unbounded "WHERE ctid > (x,y)" query.
+    // The actual last page is approximated. The last subquery will go until the end of table.
+    return chunks;
+  }
+
   private AutoCloseableIterator<RowDataWithCtid> queryTableCtid(
                                                                 final List<String> columnNames,
                                                                 final String schemaName,
-                                                                final String tableName) {
+                                                                final String tableName,
+                                                                final long tableSize,
+                                                                final long blockSize) {
 
     LOGGER.info("Queueing query for table: {}", tableName);
     final AirbyteStreamNameNamespacePair airbyteStream =
         AirbyteStreamUtils.convertFromNameAndNamespace(tableName, schemaName);
-    return AutoCloseableIterators.lazyIterator(() -> {
+
+    final CtidStatus currentCtidStatus = ctidStateManager.getCtidStatus(airbyteStream);
+
+    // Rather than trying to read an entire table with a "WHERE ctid > (0,0)" query,
+    // We are creating a list of lazy iterators each holding a subquery according to the plan.
+    // All subqueries are then composed in a single composite iterator.
+    // Because list consists of lazy iterators, the query is only executing when needed on after the other.
+    final List<Pair<Ctid, Ctid>> subQueriesPlan = ctidQueryPlan((currentCtidStatus == null) ? Ctid.of(0,0) : Ctid.of(currentCtidStatus.getCtid()), tableSize, blockSize, QUERY_TARGET_SIZE_GB);
+    final List<AutoCloseableIterator<RowDataWithCtid>> subQueriesIterators = new ArrayList<>();
+    subQueriesPlan.forEach(p -> subQueriesIterators.add(AutoCloseableIterators.lazyIterator(() -> {
       try {
         final Stream<RowDataWithCtid> stream = database.unsafeQuery(
-            connection -> createCtidQueryStatement(connection, columnNames, schemaName, tableName, airbyteStream), sourceOperations::recordWithCtid);
+            connection -> createCtidQueryStatement(connection, columnNames, schemaName, tableName, p.getLeft(), p.getRight()),sourceOperations::recordWithCtid);
+
         return AutoCloseableIterators.fromStream(stream, airbyteStream);
       } catch (final SQLException e) {
         throw new RuntimeException(e);
       }
-    }, airbyteStream);
+    }, airbyteStream)));
+    return AutoCloseableIterators.concatWithEagerClose(subQueriesIterators);
```

Comment:
nice! so this is essentially creating one iterator so that we enforce the one stream, one iterator principle?

Reply:
exactly. because from the outside it's a single iterator we just keep doing next() 
Internally we query the first, one and drain it out, then query the second and so on.
at each time there's a single open connection.

When we parallelize source-postgres it should be easy since it can be done on the outer composite iterator level.